### ISSUE TITLE
ConstantMaterial

### ISSUE THREAD

**Issue by friedmud** [idaholab/moose#6979]
### Description of the enhancement or error report

As we were discussing on #6910 having a `ConstantMaterial` would be useful.  It would compute its values _once_ into a vector that is "max qps" long... and then that would simply be used for the whole duration of the simulation.
### Rationale for the enhancement or information for reproducing the error

This would greatly speed up some calculations
### Identified impact

There would be some impact to the internals of MOOSE.  Some bookkeeping would have to be created to keep track of `ConstantMaterials` and deal with them in a special way.  In particular, there would have to be some code for swapping out `ConstantMaterials` when subdomains change.

**Comment by YaqiWang** [comment]
Can we support multiple `ConstantMaterial` which are defined on subdomains?

**Comment by friedmud** [comment]
That is the plan.

**Comment by aeslaughter** [comment]
This seems like a stateful Material with "execute_on=initial", perhaps supporting "execute_on" is a more generic approach.

If you do nothing (i.e., set "execute_on='initial linear nonlinear') you get the current behavior with on-the-fly calculation of properties. If you alter execute_on to something else, we enable stateful storage and proceed.

**Comment by permcody** [comment]
I think this is a bad idea until we can do load balancing. If you have a constant material how expensive can it be to compute? Do you really want to store a million elements worth of constant data and screw your load balancing too?

**Comment by dschwen** [comment]
I'm a little confused as to how the execute_on schema would apply to materials. Material properties are computed when they are needed. Bringing execute on into this picture would make some stateful that otherwise don't need to be stateful. The ConstantMaterial does not need to be a stateful material at all! It just needs to reset its values when the subdomain changes.

This begs another question. Do we even want constant materials having different values on different domains? This would essentially allow people to do spatially varying properties through the back door.

**Comment by aeslaughter** [comment]
I will add that doing this execute on approach should not be that difficult. It would require the following:
1. Add the 'execute_on' parameter and switch the `Material` warehouse to be a `Execute...Warehouse`.
2. Enable stateful storage when the `execute_on` flags are changed.
3. Pass in an an additional argument to `FEProblem::reinitMaterials` containing the ExecFlag.

All of this wouldn't be that difficult to implement and would allow for materials to be compute at specific times, consistent with other systems in MOOSE.

**Comment by dschwen** [comment]
But... why?

**Comment by aeslaughter** [comment]
I just see this exploding into TimestepMaterial, BoundaryConstantMaterial, BoundaryFunctionMaterial, BoundaryTimestepFunctionMaterial, SubdomainConstantMaterial.

**Comment by permcody** [comment]
It might, but I know we aren't ready for `execute_on` Materials today. We still have odd bugs here and there in the stateful system. This might be a good idea someday in the future.

**Comment by YaqiWang** [comment]
I do not think we want to use stateful material property for this. The number of constant materials is much more limited than the number of quadrature points on the mesh. It is rather that a constant material provide a function to evaluate its material properties possibly even just in `initialSetup()`. Then MOOSE will call it and store the values somewhere, which will be retrieved during `subdomainChange()`. `computeProperties()` for constant materials will be never called.

**Comment by friedmud** [comment]
Guys... this has nothing to do with stateful... and nothing to do with `execute_on`.

@YaqiWang has this exactly right.

Let's say that "max_qps" is 8.

All of the `ConstantMaterial`s will be asked to compute _8_ values up front... at the beginning of the simulation.  Those values will be stored "somewhere"... and when the subdomain changes those values will be retrieved for those material properties and swapped into the right position.  They will not change at all (not moved, not computed) until the subdomain changes again.

This could be a huge savings some physics... especially neutronics.  Neutronics requires _massive_ material properties (think 300x300 dense matrices at each quadrature point) that may be completely constant over subdomains (or even over the whole mesh!).  You don't want to have to even _copy_ those into the values for each quadrature point on each element every time!

Same can be said for really complicated `ParsedFunction` based material properties.  It's a huge waste to evaluate a `ParsedFunction` for every quadrature point on every element when the value is going to be a single constant value.

But whatevs... I put this up here for discussion... if you guys are against it go ahead and close it.

**Comment by permcody** [comment]
@friedmud - I've been on your side since the beginning. I think `execute_on` materials is a bad idea. We actually don't have this for all the other systems. I think it's more like a 50/50 split. We need to compute materials, kernels, bcs, constraints and other parts of the nonlinear system "on demand". We use `execute_on` for all auxiliary and post processing objects. When you break it into those two categories it's easy to see why the way we do Materials now, is actually the right way.

**Comment by andrsd** [comment]
I do not mind the idea of stuff being constant through out the simulation. The main work will be _probably_ in refactoring the swapping from `MaterialData` into Material, so that the `ConstantMaterial` will do its own thing, or maybe will just need `ConstantMaterialData`? Just a few thoughts on how the implementation may look like.

**Comment by andrsd** [comment]
Continuation from #6910:

> @andrsd: MaterialData is being swapped when we hit an element, so that stateful and normal material properties are behaving the same for users. So, this will be more complicated than it looks.

&nbsp;

> @friedmud: It's swapped on a per-property basis - right?

Yes, I think that is right.

**Comment by YaqiWang** [comment]
This issue is getting our attention again. We need to optimize one of our materials. We were thinking to add a parameter for `Material`, something like
```
addParam<MooseEnum>("constant_level", MooseEnum("material subdomain element none", "none"), "On what level properties in this material is constant");
```
To do this, we need MOOSE to provide a machinery to copy properties of one material from one quadrature point to another. Then we can consider to rewriting `Material::initialSetup()`, `Material::subdomainChange()`, `Material::computeProperties()`. Thanks.

**Comment by permcody** [comment]
I don't see us getting to this before the Tiger team. It's a good task to work out then though.

**Comment by friedmud** [comment]
Yep; let's talk about it in Texas

**Comment by YaqiWang** [comment]
Just let you know, the PR is not for the initial thought of @friedmud . We possibly do not want to close this issue.

**Comment by permcody** [comment]
Implemented and closed by #9456.
